src/Npgsql/NpgsqlTypes/PostgisTypes.cs

```diff
@@ -307,193 +640,332 @@ public bool Equals([CanBeNull] PostgisMultiPoint other)
-        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        public PostgisMultiPolygon(IEnumerable<IEnumerable<IEnumerable<Coordinate2D>>> ringList) =>
+            _polygons = ringList.Select(x => new PostgisPolygon(x)).ToArray();
+    }
 
-        public PostgisGeometryCollection(PostgisGeometry[] geometries)
-        {
-            _geometries = geometries;
-        }
+    /// <summary>
+    /// Represents a PostGIS 3DZ MultiPolygon.
+    /// </summary>
+    public class PostgisMultiPolygonZ : PostgisMultiPolygon<Coordinate3DZ>
+    {
+        internal override WkbIdentifier Identifier => WkbIdentifier.MultiPolygonZ;
 
-        public PostgisGeometryCollection(IEnumerable<PostgisGeometry> geometries)
-        {
-            _geometries = geometries.ToArray();
-        }
+        public PostgisMultiPolygonZ(PostgisPolygonZ[] polygons) => _polygons = polygons;
+
+        public PostgisMultiPolygonZ(IEnumerable<PostgisPolygonZ> polygons) => _polygons = polygons.ToArray();
+
+        public PostgisMultiPolygonZ(IEnumerable<IEnumerable<IEnumerable<Coordinate3DZ>>> ringList) =>
+            _polygons = ringList.Select(x => new PostgisPolygonZ(x)).ToArray();
+    }
 
-        public bool Equals([CanBeNull] PostgisGeometryCollection other)
+    /// <summary>
+    /// Represents a PostGIS 3DM MultiPolygon.
+    /// </summary>
+    public class PostgisMultiPolygonM : PostgisMultiPolygon<Coordinate3DM>
+    {
+        internal override WkbIdentifier Identifier => WkbIdentifier.MultiPolygonM;
+
+        public PostgisMultiPolygonM(PostgisPolygonM[] polygons) => _polygons = polygons;
+
+        public PostgisMultiPolygonM(IEnumerable<PostgisPolygonM> polygons) => _polygons = polygons.ToArray();
+
+        public PostgisMultiPolygonM(IEnumerable<IEnumerable<IEnumerable<Coordinate3DM>>> ringList) =>
+            _polygons = ringList.Select(x => new PostgisPolygonM(x)).ToArray();
+    }
+
+    /// <summary>
+    /// Represents a PostGIS 4D MultiPolygon.
+    /// </summary>
+    public class PostgisMultiPolygonZM : PostgisMultiPolygon<Coordinate4D>
+    {
+        internal override WkbIdentifier Identifier => WkbIdentifier.MultiPolygonZM;
+
+        public PostgisMultiPolygonZM(PostgisPolygonZM[] polygons) => _polygons = polygons;
+
+        public PostgisMultiPolygonZM(IEnumerable<PostgisPolygonZM> polygons) => _polygons = polygons.ToArray();
+
+        public PostgisMultiPolygonZM(IEnumerable<IEnumerable<IEnumerable<Coordinate4D>>> ringList) =>
+            _polygons = ringList.Select(x => new PostgisPolygonZM(x)).ToArray();
+    }
+
+    /// <summary>
+    /// Base PostGIS GeometryCollection type of specified coordinates system.
+    /// </summary>
+    /// <typeparam name="T">Coordinate type for Geometry</typeparam>
+    public abstract class PostgisGeometryCollection<T> : PostgisGeometry<T>, IEquatable<PostgisGeometryCollection<T>>, IEnumerable<PostgisGeometry<T>> where T : Coordinate2D
```

Comment:
Can you confirm that it's not possible to mix geometries of different dimensions in the same geometry collection? In other words, that it's not possible to have an XY polygon and an XYZ polygon inside the same collection instance?

Reply:
Yes, that used to be the case. With the changes to make coordinates structs, and no common interface, this cannot happen anymore.